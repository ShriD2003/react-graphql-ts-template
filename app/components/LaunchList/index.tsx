import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { get, isEmpty } from 'lodash-es';
import { Card, Skeleton } from 'antd';
import { If, T, For } from '@components';
import { colors } from '@app/themes';
import { Launch } from '@app/containers/HomeContainer';

const CustomCard = styled(Card)`
  && {
    color: ${colors.primary};
    background-color: ${colors.secondaryText};
  }
`;
const WrapperCard = styled(Card)`
  && {
    color: ${colors.primary};
    border: none;
    background-color: ${colors.secondaryText};
  }
`;

const CustomErrorCard = styled(Card)`
  && {
    color: ${colors.secondary};
    margin: 2rem;
    background-color: ${colors.secondaryText};
  }
`;

const Container = styled.div`
  && {
    display: flex;
    flex-direction: column;
    width: 100%;
    margin: 0 auto;
    background-color: ${colors.secondaryText};
  }
`;

interface LaunchListProps {
  launchData: { launches?: Launch[] };
  loading: boolean;
}

export function LaunchList({ launchData, loading }: LaunchListProps) {
  const launches = get(launchData, 'launches', []);
  return (
    <If
      condition={!isEmpty(launches) || loading}
      otherwise={
        <CustomErrorCard>
          <T data-testid="default-message" id="fallback" />
        </CustomErrorCard>
      }
    >
      <WrapperCard data-testid="list">
        <Skeleton loading={loading} active>
          <For
            of={launches}
            ParentComponent={Container}
            renderItem={(launch: Launch, idx) => (
              <CustomCard key={idx}>
                <div>{launch.mission_name}</div>
                <div> {launch.launch_date_local}</div>
              </CustomCard>
            )}
          />
        </Skeleton>
      </WrapperCard>
    </If>
  );
}

LaunchList.propTypes = {
  launchData: PropTypes.shape({
    launches: PropTypes.arrayOf(
      PropTypes.shape({
        mission_name: PropTypes.string,
        launch_date_local: PropTypes.string,
        links: PropTypes.shape({
          wikipedia: PropTypes.string,
          flick_images: PropTypes.array
        })
      })
    )
  }),
  loading: PropTypes.bool
};

export default memo(LaunchList);
