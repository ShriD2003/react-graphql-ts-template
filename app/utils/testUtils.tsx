import React from 'react';
import { I18nProvider } from '@lingui/react';
import { i18n } from '@lingui/core';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import history from './history';

export const renderWithIntl = (children: React.ReactNode) => {
  i18n.load(DEFAULT_LOCALE, translationMessages[DEFAULT_LOCALE]);
  i18n.activate(DEFAULT_LOCALE);

  return render(<I18nProvider i18n={i18n}> {children} </I18nProvider>);
};

export const getComponentStyles = (Component: React.FC<any>, props = {}) => {
  renderWithIntl(Component(props));
  const { styledComponentId } = Component(props)!.type;
  const componentRoots = document.getElementsByClassName(styledComponentId);
  return window.getComputedStyle(componentRoots[0]);
};

export const renderProvider = (children: React.ReactNode, { path }: { path?: string } = {}, renderFn = render) => {
  const store = configureStore({}).store;
  return renderFn(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider
          theme={{
            main: 'violet'
          }}
        >
          <Router history={history}>{path ? <Route path={path}>{children}</Route> : children}</Router>
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = <Data,>(ok: boolean, data: Data, error?: object) => ({
  ok,
  data,
  error
});
