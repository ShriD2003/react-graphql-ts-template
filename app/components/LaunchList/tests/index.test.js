/**
 *
 * Tests for LaunchList
 *
 */

import React from 'react';
// import { fireEvent } from '@testing-library/dom'
import { renderWithIntl } from '@utils/testUtils';
import LaunchList from '../index';

describe('<LaunchList />', () => {
  const loading = false;
  const launchData = {
    launches: [
      {
        launch_date_local: '2014-01-06T14:06:00-04:00',
        mission_name: 'Thaicom 6',
        links: {
          wikipedia: 'https://en.wikipedia.org/wiki/Thaicom_6',
          flickr_images: ['https://farm9.staticflickr.com/8617/16789019815_f99a165dc5_o.jpg']
        }
      }
    ]
  };
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<LaunchList loading={loading} launchData={launchData} />);
    expect(baseElement).toMatchSnapshot();
  });
  it('should show the fallbackMessage if the  ', () => {
    const message = 'No results found for the search term.';
    const { getByText } = renderWithIntl(<LaunchList loading={loading} launchData={null} />);
    expect(getByText(message)).toBeInTheDocument();
  });
  it('should render the list for the launches when data is available', () => {
    const { getByText } = renderWithIntl(<LaunchList loading={loading} launchData={launchData} />);
    expect(getByText('Thaicom 6')).toBeInTheDocument();
  });
});
