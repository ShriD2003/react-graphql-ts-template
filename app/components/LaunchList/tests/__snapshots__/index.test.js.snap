// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaunchList /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered LaunchList__CustomErrorCard-sc-1l6drbj-2 kKRUhr"
    >
      <div
        class="ant-card-body"
      >
        <p
          class="T__StyledText-gcrjip-0 fIysFh"
          data-testid="default-message"
        >
          No results found for the search term.
        </p>
      </div>
    </div>
  </div>
</body>
`;
